Step3

lcjltj opened the pull request:
3단계(자동) 코드 리뷰 요청 드립니다. !!

lcjltj committed:
docs: 문자열 계산기 기능 요구사항 정의

lcjltj committed:
feat: 문자열 계산기 split 기능 추가

- 사용자 정의 구분자 split
- (,), (:) 구분자 split

lcjltj committed:
feat: 문자열 계산기 split 기능 추가

- 사용자 정의 구분자 split
- (,), (:) 구분자 split

lcjltj committed:
feat(StringCalculator): 문자열 sum 기능 추가

- isEmpty() 공백, null 확인 메소드 추가
- isValid() 음수 입력시 Exception 메소드 추가
- lessThenZero() 음수 확인 메소드 추가
- sum() 합계 메소드 추가

lcjltj committed:
feat(StringCalculator): 문자열 sum 기능 추가

- isEmpty() 공백, null 확인 메소드 추가
- isValid() 음수 입력시 Exception 메소드 추가
- lessThenZero() 음수 확인 메소드 추가
- sum() 합계 메소드 추가

lcjltj committed:
refactor: 재사용 가능한 메소드 분리 및 메소드 명 변경

- isEmpty() 메소드 StringUtils 클래스 로 이동
- toNumbers() 메소드 ColltionsUtils 클래스 로 이동
- lessThenZero() 메소드 명 isLessThenZero()로 변경
- isLessThenZero() 메소드 NumberUtils 클래스 로 이동
- calculator package 추가

lcjltj committed:
refactor: 재사용 가능한 메소드 분리 및 메소드 명 변경

- isEmpty() 메소드 StringUtils 클래스 로 이동
- toNumbers() 메소드 ColltionsUtils 클래스 로 이동
- lessThenZero() 메소드 명 isLessThenZero()로 변경
- isLessThenZero() 메소드 NumberUtils 클래스 로 이동
- calculator package 추가

lcjltj committed:
docs: 로또 기능 요구 사항 정의

lcjltj committed:
docs: 로또 기능 요구 사항 정의

lcjltj committed:
feat(LottoNumber): 로또 번호 포장 클래스 생성

- validate() 1 ~ 45 값이 아닐 경우 Exception 발생

lcjltj committed:
feat(LottoNumber): 로또 번호 포장 클래스 생성

- validate() 1 ~ 45 값이 아닐 경우 Exception 발생

lcjltj committed:
feat(Lotto): 로또 클래스 생성

- Lotto() 생성자 생성시 1 ~ 45 의 로또 넘버 리스트 생성
- shuffle() 1 ~ 45 자리 로또 혼합 메소드 생성
- autoLotto() 6자리의 자동 로또 번호 생성

lcjltj committed:
feat(Lotto): 로또 클래스 생성

- Lotto() 생성자 생성시 1 ~ 45 의 로또 넘버 리스트 생성
- shuffle() 1 ~ 45 자리 로또 혼합 메소드 생성
- autoLotto() 6자리의 자동 로또 번호 생성

lcjltj committed:
feat(LottoNumber): 로또 넘버 getter 메소드 추가

lcjltj committed:
feat(LottoNumber): 로또 넘버 getter 메소드 추가

lcjltj committed:
feat(Lotto): 로또 validation 및 matches 메소드 추가

- Lotto(List<LottoNumber>) 생성자 추가
- validate() 메소드 추가 validateLength() 6자리 비교
                      validateOverlap() 중복 확인
- matches(Lotto) 당첨 로또 번호에 포함 갯수 확인
- contains(LottoNumber) 각 로또 번호 포함 유무 확인
- printLotto() 현재 로또 번호 출력 메소드 추가

lcjltj committed:
feat(Lotto): 로또 validation 및 matches 메소드 추가

- Lotto(List<LottoNumber>) 생성자 추가
- validate() 메소드 추가 validateLength() 6자리 비교
                      validateOverlap() 중복 확인
- matches(Lotto) 당첨 로또 번호에 포함 갯수 확인
- contains(LottoNumber) 각 로또 번호 포함 유무 확인
- printLotto() 현재 로또 번호 출력 메소드 추가

lcjltj committed:
feat: LottoMachine, Money 클래스 추가

- Money.class 금액 관련 포장 클래스 추가

- LottoMachine.class
  - validate(money) 금액 부족 관련 오류 메소드 추가
  - purchase(money) 입력 받은 금액 만큼 로또 구입 메소드 추가
  - calculate(money) 입력 받은 금액 으로 구입 가능한 로또 수 메소드 추가
  - generateAuto(count) 구입 가능한 로또 만큼 로또 자동 생성 메소드 추가

lcjltj committed:
feat: LottoMachine, Money 클래스 추가

- Money.class 금액 관련 포장 클래스 추가

- LottoMachine.class
  - validate(money) 금액 부족 관련 오류 메소드 추가
  - purchase(money) 입력 받은 금액 만큼 로또 구입 메소드 추가
  - calculate(money) 입력 받은 금액 으로 구입 가능한 로또 수 메소드 추가
  - generateAuto(count) 구입 가능한 로또 만큼 로또 자동 생성 메소드 추가

lcjltj committed:
refactor: LottoNumber 스태틱 배열 추가

- of(inputNumber) 로또 번호 일 경우 1 ~ 45 의 고정 값을 사용 하므로 스태틱 배열 추가 및 팩토리 메소드 패턴 추가
- 스태틱 상수 변경 MIN -> BEGIN_NUMBER , MAX -> END_NUMBER IntStream range 함수는 마지막 범위는 포함 하지 않아 변경
- matches 함수 명 변경 -> matchCount()

lcjltj committed:
refactor: LottoNumber 스태틱 배열 추가

- of(inputNumber) 로또 번호 일 경우 1 ~ 45 의 고정 값을 사용 하므로 스태틱 배열 추가 및 팩토리 메소드 패턴 추가
- 스태틱 상수 변경 MIN -> BEGIN_NUMBER , MAX -> END_NUMBER IntStream range 함수는 마지막 범위는 포함 하지 않아 변경
- matches 함수 명 변경 -> matchCount()

lcjltj committed:
feat(Prize): 상금 enum 클래스 추가

lcjltj committed:
feat(Prize): 상금 enum 클래스 추가

lcjltj committed:
feat(Prize): 상금 enum 클래스 추가

lcjltj committed:
feat(Prize): 상금 enum 클래스 추가

lcjltj committed:
feat(): Winners 클래스 추가 외 메소드 2개 추가

- Winners 당첨 금액 클래스 생성
- Prize 당첨 리스트 Winners 클래스 포장
    - calculatePrize() 당첨 수에 맞는 금액 리턴
    - print() 현재 상금 관련 내용 출력
    - get 메소드 제거

lcjltj committed:
feat(): Winners 클래스 추가 외 메소드 2개 추가

- Winners 당첨 금액 클래스 생성
- Prize 당첨 리스트 Winners 클래스 포장
    - calculatePrize() 당첨 수에 맞는 금액 리턴
    - print() 현재 상금 관련 내용 출력
    - get 메소드 제거

lcjltj committed:
feat(PrizeExchanger): 로또 상금 교환 클래스 생성

- exchange() 당첨된 로또의 전체 상금액 계산 메소드 추가

lcjltj committed:
feat(PrizeExchanger): 로또 상금 교환 클래스 생성

- exchange() 당첨된 로또의 전체 상금액 계산 메소드 추가

lcjltj committed:
refactor: 오류 문구 ExceptionConstants 분리 외 2

- exception 문구 분리
- InputView , OutputView 클래스 추가

lcjltj committed:
refactor: 오류 문구 ExceptionConstants 분리 외 2

- exception 문구 분리
- InputView , OutputView 클래스 추가

lcjltj committed:
feat(LottoController): LottoController 클래스 추가 외 1

- LottoController 클래스 추가
    - buyLotto(money) 입력 금액 만큼 로또 구입
    - exchangePrize(Lottoes, answer) 구입한 로또들과 당첨번호로 총 지급액 계산
    - printYield(totalPrize, money) 구입한 금액 당첨금액 수익율 계산
- ResultView 로또 번호 출력 메소드 추가

lcjltj committed:
feat(LottoController): LottoController 클래스 추가 외 1

- LottoController 클래스 추가
    - buyLotto(money) 입력 금액 만큼 로또 구입
    - exchangePrize(Lottoes, answer) 구입한 로또들과 당첨번호로 총 지급액 계산
    - printYield(totalPrize, money) 구입한 금액 당첨금액 수익율 계산
- ResultView 로또 번호 출력 메소드 추가

lcjltj committed:
Merge remote-tracking branch 'origin/step3' into step3

lcjltj committed:
docs: 기능 요구 사항 확인 업데이트

parkeeseul commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,112 @@
+                .map(LottoNumber::getLottoNumber)
+                .map(Objects::toString)
+                .collect(Collectors.joining(PRINT_DELIMITER, PRINT_PREFIX, PRINT_SUFFIX));
+    }
```
해당 코드는 UI 를 담당하는 로직이네요.
ResultView와 같은 UI 를 담당하는 클래스로 분리해 주면 어떨까요? 😄 
```
프로그램 요구사항
- 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.
```

lcjltj replied:
이슬님! 해당 로직 ResultView로 변경 하였습니다.
기존에 Lotto에 있었던 이유는 로또에 관한 getter를 사용하지 않으려고 했었습니다 ㅠㅠ

parkeeseul replied:
뷰에서 getter 를 사용하는 것은 괜찮다고 생각합니다.
하지만 만약 도메인의 getter 를 사용하는 것이 걸리신다면 DTO 를 활용해 볼 수도 있을 것 같아요! 😄

parkeeseul commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,112 @@
+        lotto = IntStream.range(BEGIN_NUMBER, END_NUMBER)
+                .mapToObj(LottoNumber::of)
+                .collect(Collectors.toList());
+    }
```
해당 부분은 번호를 미리 정의해 두고 재사용 하기 위함일까요? 🤔 
Lotto는 1 ~ 45 까지의 숫자 중 6개의 숫자로만 이루어 지는데요,
Lotto가 1~45까지의 모든 수를 가지는 것은 어색한 것 같은데 어떻게 생각하시나요?

lcjltj replied:
로또에 대한 범위를 제가 잘못 둔거 같아요!
로또는 7개의 번호를 가지고 있고, 로또를 자동으로 생성하거나 하는 로직을 별도로 뒀어야 할거같아 
LottoMachine 클래스에게 해당 기능들의 책임을 바꾸게 했습니다

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+        this.prize = prize;
+    }
+
+    public static Winners matchLotto(final List<Lotto> lottoes, final Lotto answer) {
```
answer를 로또 우승번호로 사용해 주신 것 같은데요,
answer 라는 네이밍 보다 우승번호라는 의미를 담는 네이밍을 해주는게 어떨까요?
조금 더 가독성이 좋아질 것 같아요 😄

lcjltj replied:
해당 네이밍명 변경 하였습니다.

parkeeseul commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,112 @@
+        return lotto.contains(lottoNumber);
+    }
+
+    private void validate(final List<LottoNumber> lottoNumbers) {
+        validateLength(lottoNumbers);
+        validateOverlap(lottoNumbers);
+    }
```
코드는 위에서 아래로 자연스럽게 읽히도록 하는 것이 가독성이 좋은데요,
코드의 위치를
Lotto() -> validate() -> validateLength() ...
와 같이 호출되는 사용되는 순서대로 배치해 보면 어떨까요? 😄

lcjltj replied:
안녕하세요 이슬님! 피드백 감사드립니다.
달아주신 피드백 하나씩 보면서 수정하고 있는데요!
이펙티브 자바책 에서도 코드는 위에서 아래로 자연스럽게 읽히는게 가독성이 좋다고 하는 부분을 본적이 있는데
기본적으로 구조를 짤때 아래와 같이 구조를 하려고 하고 있습니다.
물론 현재 클래스에서는 validate 사용하는 곳이 생성자에만 있지만 다른 메소드에서 저 메소드를 호출하게 될 경우
결국 위에서 아래로 읽을수 없는 구조기 때문에 아래와 같이 배치를 하고 있었습니다.
constructor
public static method
private static method
public method
private method
이슬님은 위와같은 구조로 짜는것에 대해서는 어떻게 생각하시나요 ?

parkeeseul replied:
[자바 컨벤션](https://myeonguni.tistory.com/1596) 을 확인해 보셔도 좋을 것 같아요 😄 
![스크린샷 2022-05-22 오전 11 17 06](https://user-images.githubusercontent.com/48339310/169675443-ba776df5-0a65-4f32-af5c-b84cf16c3c77.png)
C 라는 메서드를 A 와 B 에서 호출한다면
개인적으로는 A -> B -> C 와 같이 배치하는 편입니다.

parkeeseul commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.stream.IntStream;
+
+public class LottoMachine {
+    private static final int PRICE = 1000;
```
돈과 같은 단위는 언더바를 통해 좀 더 가독성을 높여볼 수 있을 것 같아요 😄 
```suggestion
    private static final int PRICE = 1_000;
```

lcjltj replied:
수정하였습니다.

parkeeseul commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,36 @@
+        return IntStream.range(ZERO, count)
+                .mapToObj(i -> Lotto.auto())
+                .collect(Collectors.toList());
+    }
```
LottoMachine 는 Model, View, Controller 중 어떤 목적으로 설계를 해 주신 걸까요? 😄

parkeeseul commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+import static lotto.constants.ExceptionConstants.*;
+
+public class LottoNumber {
+    private static final LottoNumber[] LOTTO_NUMBER;
```
List 가 아닌 배열을 사용하신 이유가 있으신가요? 😄

parkeeseul commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+                .toArray(LottoNumber[]::new);
+    }
+
+    private final int lottoNumber;
```
`LottoNumber.lottoNumber` VS `LottoNubmer.value`
어떻게 생각하시나요? 😄

lcjltj replied:
중복된 느낌이라 value가 더 쫗아보여요!!!

parkeeseul commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,23 @@
+import static lotto.constants.ExceptionConstants.*;
+
+public class Money {
+    private static final int ZERO = 0;
```
단순히 0 이라는 값을 나타내기 보다는  MIN 과 같이 값에 의미를 부여할 수 있는 네이밍은 어떨까요? 😄

parkeeseul commented on src/main/java/lotto/constants/ExceptionConstants.java:
```diff
@@ -0,0 +1,10 @@
+package lotto.constants;
+
+public class ExceptionConstants {
```
해당 메세지들을 따로 분리해 주신 이유가 있으신가요?
연관 있는 도메인에 해당 메세지들을 위치 시켰을 경우와 지금 처럼 따로 분리 했을 경우 어떠한 차이가 있을까요? 😄

lcjltj replied:
.. 정확히 분리한 이유는 다음 스텝에서 해당 익셉션에 관한 내용을 enum 과 별도 익셉션 클래스를 생성하여
사용하기 위해 따로 클래스로 빼놓았었습니다. ㅜ_ㅜ
현재 enumType 과 익셉션 클래스 생성하여 분리 하였습니다.

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+    FOURTH_PLACE(3, 5_000),
+    FAIL(0, 0);
+
+    private static final Prize[] prizes = {FAIL, FAIL, FAIL, FOURTH_PLACE, THIRD_PLACE, SECOND_PLACE, FIRST_PLACE};
```
만약 당첨 갯수에 따른 조건들이 달라지거나 늘어나게 된다면
해당 배열도 계속 늘어가거나 변경이 생길 것 같은데요,
matchCount 를 활용해 Prize 를 찾는 방법은 없을까요? 😄

lcjltj replied:
해당 방법을 우선 map 을 통해 수정을 하긴했습니다.
values로 filter를 걸어 찾을수 있지만 그렇게 되면 시간 복잡도가 O(n) 이 되기 때문에  처음에는 배열로 생성 했습니다.

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+        return prize * count;
+    }
+
+    public void printMatch(final long count) {
+        ResultView.resultLotto(matchCount, prize, count);
+    }
```
도메인에서 UI 관련 기능을 호출하고 있네요 🤔

lcjltj replied:
해당 로직 ResultView로 변경했습니다.

parkeeseul commented on src/main/java/lotto/domain/PrizeExchanger.java:
```diff
@@ -0,0 +1,24 @@
+            totalPrize += rank.calculatePrize(count);
+        }
+        return totalPrize;
+    }
```
해당 부분도 도메인에서 UI 관련 기능을 호출하고 있네요
개선해 볼 수 없을까요? 😄

lcjltj replied:
해당 로직 ResultView로 변경했습니다.

parkeeseul commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,53 @@
+        Lotto lottoTicket = new Lotto(lotto);
+        assertThat(lottoTicket.matchCount(new Lotto(lotto))).isEqualTo(6);
+        assertThat(lottoTicket.matchCount(new Lotto(lottoTwo))).isEqualTo(5);
+    }
```
같은 목적의 테스트를 두번 진행하신 이유가 있으신가요? 😄

lcjltj replied:
맞은 횟수가 서로 상이하여 한개 더 작성 했었습니다 ㅜㅜ.. 6개일때 6개인지 5개일때 5개인지

parkeeseul replied:
@MethodSource 를 활용해 볼 수도 있을 것 같아요 😄

parkeeseul reviewed (changes requested):
3단계 미션 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다! 😄 
즐거운 밤 되세요! 화이팅 입니다 🔥

lcjltj committed:
refactor(Prize): 배열로 선언된 Prize Map 방식으로 변경

lcjltj committed:
refactor(Lotto): UI(print)로직 제외 외 1

- Lotto class
    - printLotto() UI 관련 메소드 삭제(OutView 클래스에서 구현)
    - suffle() , generateAuto() , auto() 자동 로또 메소드 삭제(LottoMachine 클래서에서 구현)
- LottoNumber
    - 필드 명 변경 lottoNumber -> value
    - static 필드명 변경 -> LOTTO_NUMBER -> COLLECTION

lcjltj committed:
refactor(LottoNumber): LottoNumber 필드명 변경

- LottoNumber
    - 필드 명 변경 lottoNumber -> value
    - static 필드명 변경 -> LOTTO_NUMBER -> COLLECTION

lcjltj committed:
임시 저장 백업

lcjltj committed:
refactor(LottoException): 로또 에서 발생하는 ExceptionType enum 으로 변경

lcjltj committed:
refactor: 지난 주 로또 번호 parameter명 외 2

- answer -> winnigNumber
- static ZERO 변수명을 의미 있게 변경 ZERO -> PRIZE_MIN, MONEY_MIN
- 로또 프로그램에서 사용 되는 에러 throw LottoException

lcjltj committed:
refactor(LottoMachine): 로또 번호 생성 로직 이동 및 UI담당 로직 분리

lcjltj committed:
refactor: UI 관련 로직 이동

lcjltj committed:
refactor: 통계 출력문 추가

parkeeseul commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,24 @@
+import lotto.exception.LottoExceptionType;
+
+public class Money {
+    private static final int MONEY_MIN = 0;
```
`Money.MONEY_MIN` VS `Money.MIN`
어떻게 생각하시나요? 😄

lcjltj replied:
money를 붙이게 되면 중복 느낌이 강한거 같아요.
수정하겠습니다!

parkeeseul commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -0,0 +1,51 @@
+package lotto.domain;
+
+import lotto.view.ResultView;
```
불필요한 import 는 제거해 주세요 😄

lcjltj replied:
확인 한다고 했는데 ㅜㅜ 자꾸 뺴먹네요 수정했습니다.

parkeeseul commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,24 @@
+import lotto.exception.LottoException;
+import lotto.exception.LottoExceptionType;
+
+public class Money {
```
Money 에 대한 테스트 코드도 있으면 좋을 것 같아요 😄

lcjltj replied:
작성했습니다.!

parkeeseul commented on src/main/java/lotto/domain/Winners.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.*;
+
+public class Winners {
```
Winners 에 대한 테스트 코드 작성해 주면 좋을 것 같아요 😄

lcjltj replied:
작성했습니다.!

parkeeseul reviewed (changes requested):
피드백 잘 반영해 주셨네요!
추가적인 코멘트와 질문 주신 사항에 대한 저의 의견 남겨두었으니 확인 부탁드립니다 😄 
즐거운 일요일 되세요 🔥

lcjltj committed:
refactor: Money static 변수 수정

- MONEY_MIN -> MIN

lcjltj committed:
test: Winner, Lotto 테스트 작성 외 1

- LottoTest MethodSource 사용으로 테스트 변경
- WinnerTest
    get_rank_test() 각 당첨 개수 확인
    get_total_prize_test() 전체 금액 확인
- MoneyTest
   minus_money_test() 마이너스 금액 exception 확인

parkeeseul commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -22,10 +28,16 @@ void create_lotto_overlap_test() {
 
-    @Test
+    @ParameterizedTest
+    @MethodSource("lottoes")
     @DisplayName("로또 번호 맞은 횟수 테스트")
-    void lotto_matches_test() {
+    void lotto_matches_test(List<Lotto> lottoes) {
         Lotto lottoTicket = new Lotto("1, 2, 3, 4, 5, 7");
-        assertThat(lottoTicket.matchCount(new Lotto("1, 2, 3, 4, 5, 6"))).isEqualTo(5);
+        assertThat(lottoTicket.matchCount(lottoes.get(0))).isEqualTo(5L);
+        assertThat(lottoTicket.matchCount(lottoes.get(1))).isEqualTo(6L);
+    }
+
+    static Stream<List<Lotto>> lottoes() {
+        return Stream.of(Arrays.asList(new Lotto("1,2,3,4,5,6"), new Lotto("1,2,3,4,5,7")));
     }
```
다음과 같이 해볼 수도 있을 것 같아요 😄 
```
    @ParameterizedTest
    @MethodSource("lottoes")
    @DisplayName("로또 번호 맞은 횟수 테스트")
    void lotto_matches_test(Lotto lotto, int count) {
        Lotto lottoTicket = new Lotto("1, 2, 3, 4, 5, 7");
        assertThat(lottoTicket.matchCount(lotto)).isEqualTo(count);
    }

    private static Stream<Arguments> lottoes() { // argument source method
        return Stream.of(
                Arguments.of(new Lotto("1,2,3,4,5,6"), 5),
                Arguments.of(new Lotto("1,2,3,4,5,7"), 6)
        );
    }
```

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요! 👍 
소소한 코멘트 하나 남겨두었으니 확인 부탁드리고
추가적인 수정 사항은 다음 단계와 함께 반영 부탁드립니다 😄 
다음 단계 진행해 주세요! 🔥
